### PR TITLE
RL: Offload optimizer via torch_memory_saver

### PR DESCRIPTION
**Pull request by tdene**
# What does this PR do ?
The optimizer `offload_to_cpu()` / `restore_from_cpu()` methods in `MegatronOptimizer` use `.cpu()` and `.cuda()` to move tensors between devices. This destroys the original GPU storage and creates entirely new tensors on restore.

Offloading / onloading them via TMS prevents fragmentation due to repeated allocation/deallocation.

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

**Comment by ZhiyuLi-Nvidia** [comment]
Just post some relevant work in a similar RL framework

https://github.com/THUDM/slime/blob/8efb1166c8e4bb4c810059b0161cbd06bcfbb6cf/docs/en/blogs/release_v0.1.0.md?plain=1#L68-L88

File: megatron/core/optimizer/optimizer.py [hunk 1]
```diff
@@ -376,16 +369,29 @@ def offload_to_cpu(self):
...
-            for param_group in self.optimizer.param_groups:
-                for p in param_group['params']:
-                    if isinstance(p, torch.Tensor) and p.is_cuda:
-                        p.data = p.data.cpu()
 
-            for state_dict in self.optimizer.state.values():
-                for k, v in state_dict.items():
-                    if isinstance(v, torch.Tensor) and v.is_cuda:
-                        state_dict[k] = v.cpu()
+            if self.config.rl_offload_optimizer_during_inference:
```
Comment (jon-barker) [inline megatron/core/optimizer/optimizer.py]: I'm not sure about this design. I believe that optimizer.offload_to_cpu() and optimizer.restore_from_cpu() are only ever used for RL. If that's true then the real switch here would be whether or not you have torch_memory_saver available. If we _only_ want to support optimizer offloading use torch_memory_saver then we can just get rid of the if statement.

**Review by jon-barker — COMMENTED** [review]
Generally looks good. Just one design simplication i think we can make - see comment

File: megatron/core/optimizer/optimizer.py (hunk 1 above)
Comment (tdene) [reply megatron/core/optimizer/optimizer.py]: Addressed; how does it look now?

Comment (tdene) [reply megatron/core/optimizer/optimizer.py]: Fixed again.

**Review by shanmugamr1992 — APPROVED** [review]
